tsl/src/chunk_copy.c

```diff
@@ -625,6 +811,197 @@ chunk_copy(Oid chunk_relid, const char *src_node, const char *dst_node, bool del
 	}
 	PG_END_TRY();
 
-	/* Cleanup and delete the catalog entry */
-	chunk_copy_cleanup(&cc);
+	/* Finish up and delete the catalog entry */
+	chunk_copy_finish(&cc);
+}
+
+static ScanTupleResult
+chunk_copy_activity_tuple_found(TupleInfo *ti, void *const data)
+{
+	ChunkCopy **cc = data;
+
+	*cc = STRUCT_FROM_SLOT(ti->slot, ti->mctx, ChunkCopy, FormData_chunk_copy_activity);
+	return SCAN_CONTINUE;
+}
+
+static ChunkCopy *
+chunk_copy_activity_get(int32 activity_id, const char *operation_id)
+{
+	ScanKeyData scankeys[1];
+	ChunkCopy *cc = NULL;
+	int indexid;
+	MemoryContext old, mcxt;
+
+	/* Objects need to be in long lived context */
+	mcxt =
+		AllocSetContextCreate(PortalContext, "chunk copy cleanup activity", ALLOCSET_DEFAULT_SIZES);
+	old = MemoryContextSwitchTo(mcxt);
+
+	/* search via any one of the provided inputs */
+	if (activity_id > 0)
+	{
+		ScanKeyInit(&scankeys[0],
+					Anum_chunk_copy_activity_idx_id,
+					BTEqualStrategyNumber,
+					F_INT4EQ,
+					Int32GetDatum(activity_id));
+		indexid = CHUNK_COPY_ACTIVITY_PKEY_IDX;
+	}
+	else if (operation_id != NULL)
+	{
+		ScanKeyInit(&scankeys[0],
+					Anum_chunk_copy_activity_idx_operation_id,
+					BTEqualStrategyNumber,
+					F_NAMEEQ,
+					DirectFunctionCall1(namein, CStringGetDatum(operation_id)));
+		indexid = CHUNK_COPY_ACTIVITY_UKEY_OPERATION_IDX;
+	}
+	else
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid chunk copy activity identifier")));
+
+	ts_catalog_scan_one(CHUNK_COPY_ACTIVITY,
+						indexid,
+						scankeys,
+						1,
+						chunk_copy_activity_tuple_found,
+						AccessShareLock,
+						CHUNK_COPY_ACTIVITY_TABLE_NAME,
+						&cc);
+
+	/*
+	 * If a valid entry is returned then fill up the rest of the fields in the
+	 * ChunkCopy structure
+	 */
+	if (cc)
+	{
+		cc->mcxt = mcxt;
+		cc->chunk = ts_chunk_get_by_id(cc->fd.chunk_id, true);
+		cc->stage = NULL;
+
+		/* Was the chunk compressed in the meanwhile? */
+		if (ts_chunk_is_compressed(cc->chunk))
```

Comment:
Interesting point, I think this one could require testing. Not sure if this should be done here. Also we might want to hold some kind of lock/status on the chunk to avoid it maybe

Reply:
Locking will be looked at separately. But the thing is, we started a copy/move which failed. Then compression happened and then much later cleanup was called.

Reply:
I am also concerned about this case. What if someone compresses a chunk after a copy operation has failed and we still have a registered (unfinished, but not running) activity for it? Will we end up with this error then and how will someone be able to cleanup?

Also, is there any technical reason we cannot cleanup the copy activity if the chunk is compressed post failure of the activity? I assume we should still be able to cleanup the destination node since the metadata there was created prior to compression of the chunk. So, do we even need a check and error if the chunk is compressed?

Should we add a check in `compress_chunk` that there is no ongoing (and potentially failed) copy activity for the chunk before compressing?

I can imagine adding an error to `compress_chunk` that says: "cannot compress chunk due to simultaneous copy activity" or similar.

Reply:
Ok, yeah. This makes sense. On further reflection, even if the source chunk gets compressed, it should be ok to cleanup the chunk on the destination node. Removing this check.